src/backend/distributed/shared_library_init.c

```diff
@@ -2132,6 +2154,21 @@ StatisticsCollectionGucCheckHook(bool *newval, void **extra, GucSource source)
 static void
 CitusAuthHook(Port *port, int status)
 {
+	/*
+	 * Limit non-superuser client connections if citus.internal_reserved_connections
+	 * is set.
+	 */
+	if (!IsSuperuser(port->user_name) &&
+		ReservedCitusInternalBackends > 0 &&
+		!IsCitusInternalBackend() &&
+		!HaveNFreeProcs(ReservedBackends + ReservedCitusInternalBackends))
```

Comment:
ReservedBackends > 2000 on Hyperscale (Citus).

My thinking was we lower that to a small number (3) and set citus.internal_reserved_connections to a high number (e.g. 2000). In that case, the N is still the same.

This does raise the question whether we have a bunch of unnecessary overhead on HSC right now...

Reply:
Looking at the implementation of HaveNFreeProcs, its behaviour (not just its performance) is also not what we want here (and arguably is also not what superuser_reserved_connections wants either). 

Any already open internal connections will count towards the limit of the external connections.

Example:
If you have a single node with
citus.local_execution=off
max_connections=4
superuser_max_connections=0
citus.internal_reserved_connections=2

and then you open one connection with a router query, then you can only run one query at a time:
```sql
-- session 1
select * where id = 1; -- uses 1 connection for the client, and one for the internal connection (which stays cached)
-- session 2
select * where id = 2; -- client connection fails HaveNFreeProcs, since only 1 connection is still free (2 for session 1 and 1 for this client connection)
```

Reply:
true, this branch was stale, pushed the newer version now